src/hotspot/share/opto/escape.cpp

```diff
@@ -387,6 +398,318 @@ bool ConnectionGraph::compute_escape() {
+    NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. Did not find memory edge on Region.", phi->_idx);)
+    return false;
+  }
+
+  // Validate outputs:
+  //    Check if we can in fact later replace the uses of the
+  //    current Phi by Phi's of individual fields.
+  //    Conditions checked:
+  //       - The only consumers of the Phi are:
+  //           - AddP (with constant offset)
+  //           -   - Load
+  //           - Safepoint
+  //           - uncommon_trap
+  //           - DecodeN
+  //
+  // TODO: add support for other kind of users.
+  for (DUIterator_Fast imax, i = phi->fast_outs(imax); i < imax; i++) {
+    Node* use = phi->fast_out(i);
+
+    if (use->is_CallStaticJava() || use->Opcode() == Op_SafePoint) {
+      has_call_as_user = true;
+
+      if (use->is_CallStaticJava() && use->as_CallStaticJava()->is_uncommon_trap() == false) {
+        NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. Has Allocate but cannot scalar replace it. CallStaticJava is not a trap.", phi->_idx);)
+        return false;
+      }
+    }
+    else if (use->is_AddP()) {
+      if (use->in(AddPNode::Offset)->find_intptr_t_con(-1) == -1) {
+        NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. Did not find constant input for %d : AddP.", phi->_idx, use->_idx);)
+        return false;
+      }
+
+      for (DUIterator_Fast jmax, j = use->fast_outs(jmax); j < jmax; j++) {
+        Node* use_use = use->fast_out(j);
+
+        if (!use_use->is_Load()) {
+          NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. AddP use is not a Load. %d %s", phi->_idx, use_use->_idx, use_use->Name());)
+          return false;
+        }
+      }
+    }
+    else if (use->is_DecodeN()) {
+      for (DUIterator_Fast jmax, j = use->fast_outs(jmax); j < jmax; j++) {
+        Node* use_use = use->fast_out(j);
+
+        if (!use_use->is_AddP()) {
+          NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. DecodeN use is not a Load. %d %s", phi->_idx, use_use->_idx, use_use->Name());)
+          return false;
+        }
+
+        for (DUIterator_Fast kmax, k = use_use->fast_outs(kmax); k < kmax; k++) {
+          Node* use_use_use = use_use->fast_out(k);
+
+          if (!use_use_use->is_Load()) {
+            NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. DecodeN.AddP use is not a Load. %d %s", phi->_idx, use_use_use->_idx, use_use_use->Name());)
+            return false;
+          }
+        }
+      }
+    }
+    else {
+      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. Has Allocate but cannot scalar replace it. One of the uses is: %d %s", phi->_idx, use->_idx, use->Name());)
+      return false;
+    }
+  }
+
+  if (mixed_klasses && has_call_as_user) {
```

Comment:
Add comment that only the same Klass is allowed if merged allocations are referenced by SafePoint or Uncommon
trap.

Reply:
You also need to bail out `has_nonnull_nonsr_input` case here. You can have not Allocation in such case (for example, argument) which you can't reallocate during deoptimization and you don't know its actual Klass.